src/velo-fed/VeloFarmerV3.sol

```diff
@@ -0,0 +1,715 @@
+    address public constant factory = 0xF1046053aa5682b4F9a81b5481394DA16BE5FF5a;
+    ICrossDomainMessenger public constant ovmL2CrossDomainMessenger = ICrossDomainMessenger(0x4200000000000000000000000000000000000007);
+    IRouter public constant router = IRouter(0xa062aE8A9c5e11aaA026fc2670B0D65cCc8B2858);
+    IERC20 public constant DOLA = IERC20(0x8aE125E8653821E851F12A49F7765db9a9ce7384);
+    IERC20 public constant USDC = IERC20(0x7F5c764cBc14f9669B88837ca1490cCa17c31607);
+    IERC20 public constant nUSDC = IERC20(0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85); // native USDC
+    uint32 public constant MAINNET_CCTP_DOMAIN = 0;
+    IL2ERC20Bridge public bridge;
+    address public optiFed;
+    ICCTP public immutable cctp;
+
+    error OnlyChair();
+    error OnlyGov();
+    error OnlyPendingGov();
+    error OnlyGovOrGuardian();
+    error MaxSlippageTooHigh();
+    error NotEnoughTokens();
+    error LiquiditySlippageTooHigh();
+    error RestrictedToken();
+
+    constructor(
+        address gov_,
+        address chair_,
+        address l2chair_,
+        address treasury_,
+        address guardian_,
+        address bridge_,
+        address optiFed_,
+        address cctp_,
+        uint[] memory maxSlippageBps,
+        uint maxSlippageBpsLiquidity_
+        )
+    {
+        gov = gov_;
+        chair = chair_;
+        l2chair = l2chair_;
+        treasury = treasury_;
+        guardian = guardian_;
+        bridge = IL2ERC20Bridge(bridge_);
+        optiFed = optiFed_;
+        cctp = ICCTP(cctp_);
+        maxSlippageBpsDolaToUsdc = maxSlippageBps[0];
+        maxSlippageBpsUsdcToDola = maxSlippageBps[1];
+        maxSlippageBpsUsdcNativeToDola = maxSlippageBps[2];
+        maxSlippageBpsDolaToUsdcNative = maxSlippageBps[3];
+        maxSlippageBpsUsdcToUsdcNative = maxSlippageBps[4];
+        maxSlippageBpsUsdcNativeToUsdc = maxSlippageBps[5];
+        maxSlippageBpsLiquidity = maxSlippageBpsLiquidity_;
+    }
+
+    modifier onlyGov() {
+        if (msg.sender != address(ovmL2CrossDomainMessenger) ||
+            ovmL2CrossDomainMessenger.xDomainMessageSender() != gov
+        ) revert OnlyGov();
+        _;
+    }
+
+    modifier onlyPendingGov() {
+        if (msg.sender != address(ovmL2CrossDomainMessenger) ||
+            ovmL2CrossDomainMessenger.xDomainMessageSender() != pendingGov
+        ) revert OnlyPendingGov();
+        _;
+    }
+
+    modifier onlyChair() {
+        if ((msg.sender != address(ovmL2CrossDomainMessenger) ||
+            ovmL2CrossDomainMessenger.xDomainMessageSender() != chair) &&
+            msg.sender != l2chair
+        ) revert OnlyChair();
+        _;
+    }
+
+    modifier onlyGovOrGuardian() {
+        if ((msg.sender != address(ovmL2CrossDomainMessenger) ||
+            (ovmL2CrossDomainMessenger.xDomainMessageSender() != gov) &&
+             ovmL2CrossDomainMessenger.xDomainMessageSender() != guardian)
+        ) revert OnlyGovOrGuardian();
+        _;
+    }
```

Comment:
Can probably simplify these to a onlyRole(address) modifier, onlyGovOrRole(address) modifier and OnlyRole(address) error message.